src/SamplesApp/SamplesApp.UITests/TestFramework/ImageAssert.cs

```diff
@@ -15,159 +15,110 @@ namespace SamplesApp.UITests.TestFramework
+		private static void AreEqualImpl(
+			FileInfo expected,
+			Rectangle expectedRect,
+			FileInfo actual,
+			Rectangle actualRect,
+			double expectedToActualScale,
+			PixelTolerance tolerance,
+			int line)
 		{
-			Assert.AreEqual(expectedRect.Size, actualRect.Size, WithContext("Compare rects don't have the same size"));
+			if (expectedRect != FirstQuadrant && actualRect != FirstQuadrant)
+			{
+				Assert.AreEqual(expectedRect.Size, actualRect.Size, WithContext("Compare rects don't have the same size"));
+			}
 
 			using (var expectedBitmap = new Bitmap(expected.FullName))
 			using (var actualBitmap = new Bitmap(actual.FullName))
 			{
-				Assert.AreEqual(expectedBitmap.Size, actualBitmap.Size, WithContext("Screenshots don't have the same size"));
+				if (expectedRect == FirstQuadrant && actualRect == FirstQuadrant)
+				{
+					Assert.AreEqual(expectedBitmap.Size, actualBitmap.Size, WithContext("Screenshots don't have the same size"));
+				}
 
-				var width = Math.Min(expectedRect.Width, expectedBitmap.Size.Width);
-				var height = Math.Min(expectedRect.Height, expectedBitmap.Size.Height);
+				expectedRect = Normalize(expectedRect, expectedBitmap.Size);
+				actualRect = Normalize(actualRect, actualBitmap.Size);
 
-				(int x, int y) expectedOffset = (
-					expectedRect.X < 0 ? expectedBitmap.Size.Width + expectedRect.X : expectedRect.X,
-					expectedRect.Y < 0 ? expectedBitmap.Size.Height + expectedRect.Y : expectedRect.Y
-				);
-				(int x, int y) actualOffset = (
-					actualRect.X < 0 ? actualBitmap.Size.Width + actualRect.X : actualRect.X,
-					actualRect.Y < 0 ? actualBitmap.Size.Height + actualRect.Y : actualRect.Y
-				);
+				var expectedPixels = ExpectedPixels
+					.At(actualRect.Location)
+					.Pixels(expectedBitmap, expectedRect)
+					.Named(expected.Name)
+					.WithTolerance(tolerance);
 
-				for (var x = 0; x < width; x++)
-					for (var y = 0; y < height; y++)
-					{
-						var expectedPixel = expectedBitmap.GetPixel(x + expectedOffset.x, y + expectedOffset.y);
-						var actualPixel = actualBitmap.GetPixel(x + actualOffset.x, y + actualOffset.y);
-						if (expectedPixel == actualPixel)
-						{
-							continue;
-						}
-
-						var cumulativeError = GetColorDistance(expectedPixel, actualPixel);
-						if (cumulativeError > permittedPixelError)
-						{
-							Assert.Fail(WithContext(builder: builder => builder
-								.AppendLine($"Difference between pixels at ({x},{y}) exceeds permitted error of {permittedPixelError}")
-								.AppendLine($"expectedPixel: {ToArgbCode(expectedPixel)} {expectedPixel}")
-								.AppendLine($"actualPixel  : {ToArgbCode(actualPixel)} {actualPixel}")
-								.AppendLine($"permittedPixelError: {permittedPixelError}")
-								.AppendLine($"cumulativeError    : {cumulativeError}")
-							));
-						}
-					}
+				var report = GetContext();
+				if (Validate(expectedPixels, actualBitmap, expectedToActualScale, report))
+				{
+					Console.WriteLine(report.ToString());
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this logging statement.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5363624)